noise_corrector.py

```diff
@@ -71,149 +216,238 @@ def estimate_true_values(observed_values, randomized_response_rate, beta=0):
+  default_value: a default / null value
+
+  The algorithm used in this method attempts to achieve two goals:
+  1. The output values match the distribution as much as possible, minimizing
+     bias.
+  2. The output maximizes 'overlap' with the input. That is, satisfying (1), we
+     want to minimize the number of inputs that are adjusted.
+  '''
+
+  # Split the distribution into two chunks: elements with mass >= 1 and not.
+  # Clamp negatives to 0. Note this introduces bias, but it seems unavoidable.
+  large_elts = {k: v for k, v in estimated_counts.items() if v >= 1}
+  small_elts = {k: max(0, v) for k, v in estimated_counts.items() if v < 1}
+  assert len(small_elts) + len(large_elts) == len(estimated_counts)
+
+  def decrement_large_elt(elt):
+    large_elts[elt] -= 1
+    if large_elts[elt] < 1:
+      small_elts[elt] = large_elts[elt]
+      large_elts.pop(elt)
+
+  def handle_value(val):
+    assert val in estimated_counts and (val in small_elts or val in large_elts)
+    # Do not adjust s if there is mass in the aggregate distribution for it.
+    if val in large_elts:
+      decrement_large_elt(val)
+      return val
+
+    # Otherwise, pick from among the elements with enough mass for subtraction.
+    if len(large_elts) > 0:
+      elt = random.choices(list(large_elts.keys()), large_elts.values())[0]
+      decrement_large_elt(elt)
+      return elt
+
+    # Otherwise, sample with replacement from the small elts, with any leftover
+    # mass defaulting to a default value / null value. This also introduces
+    # small bias towards the null value bucket.
+    total_mass_remaining = sum(small_elts.values())
+    if random.random() <= total_mass_remaining:
+      return random.choices(list(small_elts.keys()), small_elts.values())[0]
+    return default_value
+
+  return [handle_value(v) for v in values]
+
+
+def generate_corrected_event_level(joined: JoinedList, params: ParamConfig) -> JoinedList:
+  # Shuffle the input data. This is because the order of which values are
+  # considered in the loop below matters, because we adjust the distribution
+  # as we go. Shuffling first avoids favoring some values over others.
+  # Note: ideally adjust_to_match_distribution would shuffle, but it becomes
+  # more difficult to associate the adjustments with the input JSON.
+  random.shuffle(joined)
+  aggregates = get_raw_corrected_map(joined, params)
+  outputs = [OutputEnumeration.create_from_data(j) for j in joined]
+  adjusted_values = adjust_to_match_distribution(
+      outputs, aggregates, OutputEnumeration.create_null(params))
+
+  def adjust_json(j, adjusted) -> Joined:
+    source: Source = j[0]
+    adjusted_reports: List[Report] = adjusted.generate_reports_for_source(
+        source)
+    return (source, adjusted_reports)
+
+  return [adjust_json(j, a) for j, a in zip(joined, adjusted_values)]
+
+
+def main():
+  DESCRIPTION = '''\
+  noise_corrector.py is a command-line tool that attempts to debias the noise
```

Comment:
Change suggestion:
Could we unify the language and always say "debias" ? Or "correct"?
I find "correct" easier to understand (but maybe "debias" is more accurate).

Reply:
Went with "debiased" since I think it's more accurate.

Reply:
Sounds good; it would be ideal to reflect this other places as well for clarity. Nothing blocking for this PR though.

Reply:
ack